### ISSUE TITLE
Allow inputs for brush size instead of capping at 200.

### ISSUE THREAD

**Issue by TheConceptBoy** [Zylann/godot_heightmap_plugin#293]
I noticed when working on very large terrains, when the global shapes are defined, the brush cap size of 200 makes terrain sculpting very long. It would be nice to have the maximum size as a separate text input where we can define our own maximum size and keep the slider for adjusting between 0 and maximum size.

Perhaps, as long as we're going this way, maybe even allow us the option to define minimum and maximum brush size values and have the slider adjust between the two.

**Comment by Zylann** [comment]
There is a reason it is capped to 200. It is because above, it can be very slow depending on which hardware you have. There could be a text box to choose an exact size, and the maximum could be increased, but having text boxes to choose the minimum and maximum without boundaries sounds a bit pointless IMO.

**Comment by TheConceptBoy** [comment]
I have a pretty decent rig that only sees a small performance dip on the 200 scale. The problem is - bad performance or not - using a 200 scale brush to blocking out large terrain shapes still takes too long. So for those who don't have a decent PC to handle painting larger than 200 brush sizes can simply not set their brushes to that high. They're in control of the values they use.

Now in terms of setting minimum and maximum being pointless - is meant to address that if you set your maximum range to something like 500 then the smallest movement of the size slider will mean a pretty drastic change in brush sizes. Allow you to set minimum and maximum can help get more precision our of the slider when the brush max size is uncapped and set to a large value. - Heck - even with the size of 200, I already have trouble fine tuning the size of the brush at lower values. And I have to do it quite a lot considering that the brush size is static and currently does not support pen pressure from graphics tablets (as noted in my other feature suggestion thread.)   So When I'm painting texture or sculpting terrain, especially around objects like buildings, I have to constantly tune brush sizes to get a natural result. Anyone doing 2D digital art and painting as well as 3D sculpting will vouch to having precise control over the brush size changes to be integral. And terrain painting / sculpting, just like 3D sculpting is no exception.

**Comment by Zylann** [comment]
I totally understand you want the maximum to be higher, and that could be changed. I just dont believe it has to be a textbox always shown in an already cramped area of the editor.

> Allow you to set minimum and maximum can help get more precision our of the slider when the brush max size is uncapped and set to a large value. - Heck - even with the size of 200, I already have trouble fine tuning the size of the brush at lower values.

This is why I think it makes more sense to have a text box to *set the size* directly, not set the maximum size.

As to pen pressure: this is specific to users using that kind of input. I think the plugin should allow specifying the min and max corresponding to various pressures, but this wont make any sense for people not using that. Also it is likely part of another issue: #245, and #294 which you opened (there are two of them now, I think one of them should be closed cuz now referring to it is scattered).

So basically the present issue is really just about being able to set larger brush sizes, and being able to choose one more precisely. I just dont necessarily agree with the solution to it though, the easiest I think is to just raise the hardcoded maximum and have a textbox to enter a specific size.

**Comment by TheConceptBoy** [comment]
> This is why I think it makes more sense to have a text box to set the size directly, not set the maximum size.
I wouldn't discard the slider too hastily. There is a very good reason why every painting application worth it's salt opts in for a slider. 


![image](https://user-images.githubusercontent.com/52581279/147845694-9630a52d-0cc5-475c-ba0b-6b8b558482e5.png)

![image](https://user-images.githubusercontent.com/52581279/147845711-89392038-9fa2-44a0-9bb6-e58ac9a81dac.png)

![image](https://user-images.githubusercontent.com/52581279/147845721-c0f6485e-b17f-4f1e-8480-08a47a26022e.png)

You can adjust the slider with a the same hand you are painting with very easily in one swift motion. If you replace the slider with a textbox, user now needs to involve the other hand or detract their attention from the screen or even look at the keyboard to type in the new value. It's also worth noting that painting applications also have a very quick way of changing the brush size with a combination of holding a shift key + left mouse button press and dragging mouse left or right to increase and decrease the size. Albeit the gesture or shortcut varies from program to program, they all offer an non distracting way of changing the brush size without even moving your mouse / pen far from where you're painting. Krita for example also has [ and ] keys to change brush size as well. Not to mention both Blender sculpting and sculpting too offer a slider AND an ability to click on the value and set your own without limits.

Godot's default UI for text inputs and buttons is very bloated, which is why I see your reason for not waning to put too many of these options in. But you can also create your own custom style override that uses much less space. It's a tool after all - versatile feature should be above making the tool look slick with fewer buttons and sliders.

![image](https://user-images.githubusercontent.com/52581279/147845763-c290a7ac-deb0-47a5-a1e4-39328415c22f.png)

**Comment by Zylann** [comment]
> I wouldn't discard the slider too hastily

I don't want to remove the slider, just want to add the textbox to it somehow. Maybe like the EditorSpinSliders of the inspector.

**Comment by TheConceptBoy** [comment]
> I don't want to remove the slider, just want to add the textbox to it somehow. Maybe like the EditorSpinSliders of the inspector.
Oh yes, so a number input. Guess we're both talking about the same thing. Indeed.

**Comment by Zylann** [comment]
Keep in mind when I say "max size", I don't mean the size you get by pressing the pen to full strength on your tablet: this is the *current size*, the size of the brush, I dunno how else to call it.

Pressure would be a modifier that applies *on top of it*, for example as a multiplier that shrinks it, which gives you the "final size", and only happens if the used hardware provides pressure data. 
It can start as being a full range multiplier: if you have brush size 200, pressure 0 will give you 0, pressure 50% gives you 100, pressure 100% will give you 200. But there could be an option to control that falloff, so that maybe it would range from 100 to 200 instead of 0 to 200. And it could have a curve instead of being linear.

I'm not a tablet user so I guess you'll have to check the result yourself once I implement it (do not expect the first version to be a full-blown Krita brush though, I have limited time^^").

On the other hand, what I meant by "max value" in earlier posts, is the maximum value the brush can possibly have, the higher end of the slider of the settings.
![image](https://user-images.githubusercontent.com/1311555/147858072-ce7b49ea-17ec-43af-a7c0-c71204a617bb.png)
See how Blender doesnt let you go above 500: THAT, is the max value. It can be exceeded if a number is typed at your own risk.
That limit may be increased, but it still exists for practical reasons: performance, so that you dont accidentally freeze the program after sliding, and UX reasons, so that the slider has a minimum and maximum to exist in the first place.
